### PR TITLE
Charmhub bundle channel

### PR DESCRIPTION
**Pull request by SimonRichardson**
When re-deploying a bundle again, against the same applications,
essentially performing an upgrade we should also take into account the
channel.

Previously channel was only viewed when deploying the first time around. 
The following also takes this into account when re-deploying. It will warn
you to use force in certain situations to ensure that you're doing
something that isn't exactly supported (swapping channels etc).

Most of the code here isn't actually complicated, it wired in the new
library updates from bundlechanges. The code essentially does another
request to the store when it notices an update to ensure that we know
what revision we can move to. All this code is super accessible and
because of the reusability of resolve charm it makes it nice and easy to
call when performing a bundle upgrade.

----

There are some potential optimisations that we could consider, but
for the first pass, I believe we can ignore it for now.

 1. Deploying the same bundle shouldn't upload a charm if we already have it. Unfortunately, this occurs because of the two-step nature of adding charms in bundle changes, where we don't have enough information at the time. This can be optimized later if required.
 2. Requesting the same charm for multiple applications could be cached per bundle resolving. Again, I'm unsure how much this would be required, but we can optimize if it becomes a problem.

## QA steps

The test needs to test quite a few scenarios:

```sh
$ juju bootstrap lxd test --no-gui
```

#### Charmstore Test

```yaml
series: bionic
applications:
  ubuntu-juju-qa:
    charm: cs:ubuntu
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
```

```sh
$ juju add-model cs-test-1
$ juju deploy ./bundle.bundle
$ juju status
Model      Controller  Cloud/Region  Version    SLA          Timestamp
cs-test-1  test        lxd/default   2.9-rc6.1  unsupported  12:24:17Z

App             Version  Status  Scale  Charm   Store       Channel  Rev  OS      Message
ubuntu-juju-qa  18.04    active      1  ubuntu  charmstore  stable    17  ubuntu  ready

Unit               Workload  Agent  Machine  Public address  Ports  Message
ubuntu-juju-qa/0*  active    idle   0        10.132.183.2           ready

Machine  State    DNS           Inst id        Series  AZ  Message
0        started  10.132.183.2  juju-d689fd-0  bionic      Running
```

With a channel that is the same as the default "" channel.

```yaml
series: bionic
applications:
  ubuntu-juju-qa:
    charm: cs:ubuntu
    channel: stable
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
```

```sh
$ juju deploy ./bundle.bundle
$ juju status
Model      Controller  Cloud/Region  Version    SLA          Timestamp
cs-test-1  test        lxd/default   2.9-rc6.1  unsupported  12:24:17Z

App             Version  Status  Scale  Charm   Store       Channel  Rev  OS      Message
ubuntu-juju-qa  18.04    active      1  ubuntu  charmstore  stable    17  ubuntu  ready

Unit               Workload  Agent  Machine  Public address  Ports  Message
ubuntu-juju-qa/0*  active    idle   0        10.132.183.2           ready

Machine  State    DNS           Inst id        Series  AZ  Message
0        started  10.132.183.2  juju-d689fd-0  bionic      Running
```

```yaml
series: bionic
applications:
  ubuntu-juju-qa:
    charm: cs:ubuntu
    channel: edge
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
```

Stable and edge channels point to the same revision, so no update is performed.

```sh
$ juju deploy ./bundle.bundle
ERROR cannot deploy bundle: upgrades not supported across channels (existing: "stable", requested: "edge"); use --force to override
$ juju deploy ./bundle.bundle --force
Located charm "ubuntu" in charm-store
Executing changes:
- upload charm ubuntu from charm-store for series bionic from channel edge with architecture=amd64
Deploy of bundle completed.
$ juju status
Model      Controller  Cloud/Region  Version    SLA          Timestamp
cs-test-1  test        lxd/default   2.9-rc6.1  unsupported  12:24:17Z

App             Version  Status  Scale  Charm   Store       Channel  Rev  OS      Message
ubuntu-juju-qa  18.04    active      1  ubuntu  charmstore  stable    17  ubuntu  ready

Unit               Workload  Agent  Machine  Public address  Ports  Message
ubuntu-juju-qa/0*  active    idle   0        10.132.183.2           ready

Machine  State    DNS           Inst id        Series  AZ  Message
0        started  10.132.183.2  juju-d689fd-0  bionic      Running
```

#### Charmhub Test

```yaml
series: bionic
applications:
  ubuntu-juju-qa:
    charm: ubuntu-juju-qa
    channel: stable
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
  ubuntu-juju-qa2:
    charm: ubuntu-juju-qa
    num_units: 1
    to:
    - "1"
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
  "1":
    constraints: arch=amd64
```

```sh
$ juju add-model one  --config charmhub-url="https://api.staging.snapcraft.io"

$ juju deploy ./bundle.bundle
Located charm "ubuntu-juju-qa" in charm-hub
Located charm "ubuntu-juju-qa" in charm-hub
Executing changes:
- upload charm ubuntu-juju-qa from charm-hub for series bionic from channel stable with architecture=amd64
- deploy application ubuntu-juju-qa from charm-hub on bionic with stable
- upload charm ubuntu-juju-qa from charm-hub for series bionic with architecture=amd64
- deploy application ubuntu-juju-qa2 from charm-hub on bionic using ubuntu-juju-qa
- add new machine 0
- add new machine 1
- add unit ubuntu-juju-qa/0 to new machine 0
- add unit ubuntu-juju-qa2/0 to new machine 1
Deploy of bundle completed.

$ juju status
Model  Controller  Cloud/Region  Version    SLA          Timestamp
one    test        lxd/default   2.9-rc6.1  unsupported  12:14:28Z

App              Version  Status   Scale  Charm           Store     Channel  Rev  OS      Message
ubuntu-juju-qa            waiting    0/1  ubuntu-juju-qa  charmhub  stable    14  ubuntu  waiting for machine
ubuntu-juju-qa2           waiting    0/1  ubuntu-juju-qa  charmhub  stable    14  ubuntu  waiting for machine

Unit               Workload  Agent       Machine  Public address  Ports  Message
ubuntu-juju-qa2/0  waiting   allocating  1        10.132.183.195         waiting for machine
ubuntu-juju-qa/0   waiting   allocating  0        10.132.183.124         waiting for machine

Machine  State    DNS             Inst id        Series  AZ  Message
0        pending  10.132.183.124  juju-e2ceb5-0  bionic      Running
1        pending  10.132.183.195  juju-e2ceb5-1  bionic      Running
```

Edit the bundle to the following:

```yaml
series: bionic
applications:
  ubuntu-juju-qa:
    charm: ubuntu-juju-qa
    channel: edge
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
  ubuntu-juju-qa2:
    charm: ubuntu-juju-qa
    num_units: 1
    to:
    - "1"
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
  "1":
    constraints: arch=amd64
```

```sh
$ juju deploy ./bundle.bundle
ERROR cannot deploy bundle: upgrades not supported across channels (existing: "stable", requested: "edge"); use --force to override
$ juju deploy ./bundle.bundle --force
Executing changes:
- upload charm ubuntu-juju-qa from charm-hub for series bionic from channel edge with architecture=amd64
- upgrade ubuntu-juju-qa from charm-hub using charm ubuntu-juju-qa for series bionic from channel edge
- upload charm ubuntu-juju-qa from charm-hub for series bionic with architecture=amd64
Deploy of bundle completed.
$ juju status
Model  Controller  Cloud/Region  Version    SLA          Timestamp
one    test        lxd/default   2.9-rc6.1  unsupported  12:16:25Z

App              Version  Status  Scale  Charm           Store     Channel  Rev  OS      Message
ubuntu-juju-qa            active      1  ubuntu-juju-qa  charmhub  edge      17  ubuntu  hello
ubuntu-juju-qa2           active      1  ubuntu-juju-qa  charmhub  stable    14  ubuntu  hello

Unit                Workload  Agent  Machine  Public address  Ports  Message
ubuntu-juju-qa2/0*  active    idle   1        10.132.183.195         hello
ubuntu-juju-qa/0*   active    idle   0        10.132.183.124         hello

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.132.183.124  juju-e2ceb5-0  bionic      Running
1        started  10.132.183.195  juju-e2ceb5-1  bionic      Running
```

Again to a different track:

```yaml
series: bionic
applications:
  ubuntu-juju-qa:
    charm: ubuntu-juju-qa
    channel: 2.0/edge
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
  ubuntu-juju-qa2:
    charm: ubuntu-juju-qa
    num_units: 1
    to:
    - "1"
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
  "1":
    constraints: arch=amd64
```

```sh
$ juju deploy ./bundle.bundle
ERROR cannot deploy bundle: upgrades not supported across channels (existing: "edge", requested: "2.0/edge"); use --force to override
$ juju deploy ./bundle.bundle --force
Executing changes:
- upload charm ubuntu-juju-qa from charm-hub for series bionic from channel 2.0/edge with architecture=amd64
- upgrade ubuntu-juju-qa from charm-hub using charm ubuntu-juju-qa for series bionic from channel 2.0/edge
  added resource foo-file
- upload charm ubuntu-juju-qa from charm-hub for series bionic with architecture=amd64
Deploy of bundle completed.
$ juju status
Model  Controller  Cloud/Region  Version    SLA          Timestamp
one    test        lxd/default   2.9-rc6.1  unsupported  12:17:53Z

App              Version  Status  Scale  Charm           Store     Channel   Rev  OS      Message
ubuntu-juju-qa            active      1  ubuntu-juju-qa  charmhub  2.0/edge   19  ubuntu  hello
ubuntu-juju-qa2           active      1  ubuntu-juju-qa  charmhub  stable     14  ubuntu  hello

Unit                Workload  Agent  Machine  Public address  Ports  Message
ubuntu-juju-qa2/0*  active    idle   1        10.132.183.195         hello
ubuntu-juju-qa/0*   active    idle   0        10.132.183.124         hello

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.132.183.124  juju-e2ceb5-0  bionic      Running
1        started  10.132.183.195  juju-e2ceb5-1  bionic      Running
```

Deploying the same thing again does not upgrade the charm.

```sh
$ juju deploy ./bundle.bundle
Executing changes:
- upload charm ubuntu-juju-qa from charm-hub for series bionic from channel 2.0/edge with architecture=amd64
- upload charm ubuntu-juju-qa from charm-hub for series bionic from channel stable with architecture=amd64
Deploy of bundle completed.
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1913631

### CODE REVIEW

File: cmd/juju/application/deployer/bundlehandler.go [hunk 1]
```diff
@@ -48,6 +48,13 @@ import (
 	"github.com/juju/juju/storage"
 )
 
+// localCharmChannel is used to identify a standard channel that all local
+// charms will adhere to. THis isn't actually required, but it makes things a
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:52]: s/THis/This/

File: cmd/juju/application/deployer/bundlehandler.go [hunk 2]
```diff
@@ -48,6 +48,13 @@ import (
...
+// lot simpler when using a selecting algorithm based on channels for charms.
+var localCharmChannel = corecharm.Channel{
+	Risk: corecharm.Risk("stable"),
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:55]: Could this (the stable risk value) perhaps be something exported by `corecharm`?

File: cmd/juju/application/deployer/bundlehandler.go [hunk 3]
```diff
@@ -380,6 +377,76 @@ func (h *bundleHandler) resolveCharmsAndEndpoints() error {
 	return nil
 }
 
+func (h *bundleHandler) resolveCharmChannelAndRevision(charmURL, charmSeries, charmChannel, arch string) (string, int, error) {
+	if h.isLocalCharm(charmURL) {
+		return charmChannel, -1, nil
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:382]: If this is a charm that I downloaded for offline deployments won't it have a revision file injected? Does it make sense to pull the revision (if present) from there?

File: cmd/juju/application/deployer/bundlehandler.go [hunk 4]
```diff
@@ -380,6 +377,76 @@ func (h *bundleHandler) resolveCharmsAndEndpoints() error {
...
+	}
+	// If the charm URL already contains a revision, return that before
+	// attempting to resolve a revision from any charm store. We can ignore the
+	// error here, as we want to just parse out the charm URL.
+	// Resolution and validation of the charm URL happens further down.
+	curl, err := charm.ParseURL(charmURL)
+	if err == nil {
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:389]: Minor nit; `curl` can be scoped as it is not used outside the if block.

```suggestion
	if curl, err := charm.ParseURL(charmURL); err == nil {
```

File: cmd/juju/application/deployer/bundlehandler.go [hunk 5]
```diff
@@ -380,6 +377,76 @@ func (h *bundleHandler) resolveCharmsAndEndpoints() error {
...
+
+// constructChannelAndOrigin attempts to construct a fully qualified channel
+// along with an origin that matches the hardware constraints and the charm url
+// source.
+func (h *bundleHandler) constructChannelAndOrigin(curl *charm.URL, charmSeries, charmChannel string, cons constraints.Value) (corecharm.Channel, commoncharm.Origin, error) {
+	var channel corecharm.Channel
+	if charmChannel != "" {
+		var err error
+		channel, err = corecharm.ParseChannelNormalize(charmChannel)
+		if err != nil {
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:433]: Minor

```suggestion
		if channel, err = corecharm.ParseChannelNormalize(charmChannel); err != nil {
```

File: cmd/juju/application/deployer/bundlehandler.go [hunk 6]
```diff
@@ -796,15 +867,15 @@ func (h *bundleHandler) addApplication(change *bundlechanges.AddApplicationChang
...
+		if origin.Track != nil {
+			track = *origin.Track
 		}
 		platform, err := utils.DeducePlatform(cons, series, h.modelConstraints)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		// A channel is needed whether the risk is valid or not.
-		channel, _ := corecharm.MakeChannel(track, h.origin.Risk, "")
+		channel, _ := corecharm.MakeChannel(track, origin.Risk, "")
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:878]: Will we get an invalid `channel` value if this returns an error? If so, we should handle the error.

File: go.mod [hunk 7]
```diff
@@ -166,3 +166,5 @@ replace (
 )
 
 replace github.com/canonical/pebble v0.0.0 => github.com/hpidcock/pebble v0.0.0-20210216034042-e91a98784399
+
+replace github.com/juju/bundlechanges/v4 => /home/simon/go/src/github.com/juju/bundlechanges
```
Comment (achilleasa) [inline go.mod:170]: Gotcha!

**Review by achilleasa — CHANGES_REQUESTED** [review]
Some minor nits and a package override that needs to be removed. Other than that it LGTM

File: cmd/juju/application/deployer/bundlehandler.go (hunk 6 above)
Comment (SimonRichardson) [reply cmd/juju/application/deployer/bundlehandler.go:869]: No, the `origin.Risk` has already been resolved much higher, see `getOrigin` and that way we should be safe to assume that we don't need to handle the error twice.

File: cmd/juju/application/deployer/bundlehandler.go (hunk 3 above)
Comment (SimonRichardson) [reply cmd/juju/application/deployer/bundlehandler.go:375]: Let me check this, but I believe that we lost the original path here when we created the bundle data and so we can't go and inspect the revision file.

**Review by achilleasa — APPROVED** [review]
LGTM

**Comment by achilleasa** [comment]
QA steps worked as expected.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**